### DISCUSSION TITLE
When using sbt, is there a way to define sub projects as intransitive dependencies in a multi-project build?

### DISCUSSION THREAD

**Discussion by tahakonuk** [scalacenter/bloop#1437] [Q&A] [answered]
Hi,
I'm trying to create a layered project structure by defining multiple projects in my `build.sbt` as below;

``` sbt
lazy val root = (project in file("."))
  .aggregate(
    model,
    service,
    gui
  )
  .settings(...)
  )

lazy val model = (project in file("model"))
  .settings(...)

lazy val service = (project in file("service"))
  .settings(...)
  .dependsOn(model)

lazy val gui = (project in file("gui"))
  .settings(...)
  .dependsOn(service)
```

And I want to be able to hide `model` from `gui`. I asked [this question in sbt repository](https://github.com/sbt/sbt/discussions/6236) and was advised that I should update the `gui` definition as follows;
```sbt
lazy val gui = (project in file("gui"))
  .dependsOn(service)
  .settings(
    Compile / internalDependencyClasspath --= (model / Compile / exportedProducts).value
  )
```
I'm using metals in visual studio code. And this works if I compile the `root` from sbt or I choose sbt as a build server in metals. But it doesn't work if I try it with bloop. Is there a way to make it work with bloop?

The way I'm testing this is that I'm using a class from `model` in `gui`. Compilation fails with sbt but passes with bloop.

**Comment by tgodzik** [comment, marked as answer]
I think this might be too specific to how sbt works, which is why when Bloop's sbt plugin extracts the structure it does not exclude the `model` subproject. This might be fixable in the plugin itself, but not a 100% sure.

**Reply by tahakonuk** [reply]
Thank you for your response. Is there another way to achieve isolation other than that the one I've shown above? I mean does sbt plugin support that sort of behaviour at all?

**Reply by tgodzik** [reply]
From what I've seen this is just not really possible currently. And since it is a bit custom workaround, it might also not interact well with IDEs.

**Reply by tahakonuk** [reply]
OK then, thanks again.

**Comment by Arthurm1** [comment]
@tahakonuk Isn't this what SBT's `Optional` configuration is for?

e.g.
```
lazy val service = (project in file("service"))
  .dependsOn(model % Optional)

lazy val gui = (project in file("gui"))
  .dependsOn(service)
```

Although I'm not that familiar with SBT and looking through the Bloop Plugin code it seems that for some reason `Optional` is not a supported configuration so the classpaths are correct but `service` lacks a _project_ dependency on `model` so ordering of compile will go wrong.
@tgodzik can you think of why `Optional` isn't supported?

There's also the `Provided` scope (which is supported by Bloop plugin) if you don't want to include `model` on `gui` runtime classpath either.  But since the plugin doesn't yet export SBT runtime classpath it doesn't really matter.

Explanation of `Provided` and `Optional` scopes are [here](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html).

**Reply by tgodzik** [reply]
I can't figure out why optional was not supported, but I think https://github.com/scalacenter/bloop/pull/1540 fixes the issue properly.

**Reply by tahakonuk** [reply]
@Arthurm1 I wasn't aware of the `Optional`. Thank you for bringing that to my attention. I've just read the docs about it and I think you are right. It can be done using `Optional`.
